notification-extension/ITBNotificationServiceExtension.swift

```diff
@@ -16,129 +16,138 @@ import UserNotifications
+                return
             }
             
-            guard let messageId = itblDictionary[JsonKey.Payload.messageId] as? String else {
-                return ""
-            }
+            let attachmentId = UUID().uuidString + ITBNotificationServiceExtension.getAttachmentIdSuffix(response: response, responseUrl: responseUrl)
+            let tempFileUrl = FileManager.default.temporaryDirectory.appendingPathComponent(attachmentId)
             
-            var actionButtons: [[AnyHashable: Any]] = []
-            if let actionButtonsFromITBLPayload = itblDictionary[JsonKey.Payload.actionButtons] as? [[AnyHashable: Any]] {
-                actionButtons = actionButtonsFromITBLPayload
-            } else {
-                #if DEBUG
-                    if let actionButtonsFromUserInfo = content.userInfo[JsonKey.Payload.actionButtons] as? [[AnyHashable: Any]] {
-                        actionButtons = actionButtonsFromUserInfo
-                    }
-                #endif
-            }
+            var attachment: UNNotificationAttachment?
             
-            var notificationActions = [UNNotificationAction]()
-            for actionButton in actionButtons {
-                if let notificationAction = createNotificationActionButton(buttonDictionary: actionButton) {
-                    notificationActions.append(notificationAction)
-                }
+            do {
+                try FileManager.default.moveItem(at: location, to: tempFileUrl)
+                attachment = try UNNotificationAttachment(identifier: attachmentId, url: tempFileUrl, options: nil)
+            } catch {
+                onFailureHandler()
+                return
             }
             
-            messageCategory = UNNotificationCategory(identifier: messageId, actions: notificationActions, intentIdentifiers: [], options: [])
-            if let messageCategory = messageCategory {
-                UNUserNotificationCenter.current().getNotificationCategories { categories in
-                    var newCategories = categories
-                    newCategories.insert(messageCategory)
-                    UNUserNotificationCenter.current().setNotificationCategories(newCategories)
-                }
+            if let attachment = attachment, let content = strongSelf.bestAttemptContent, let handler = strongSelf.contentHandler {
+                content.attachments.append(attachment)
+                handler(content)
+            } else {
+                onFailureHandler()
+                return
             }
-            
-            return messageId
-        } else {
-            return content.categoryIdentifier
         }
     }
     
-    private func createNotificationActionButton(buttonDictionary: [AnyHashable: Any]) -> UNNotificationAction? {
-        guard let identifier = buttonDictionary[JsonKey.ActionButton.identifier] as? String else {
-            return nil
+    private func lastAttemptCalled() {
```

Comment:
It is not very clear from the method name what it does. Is this method something that needs to be always called at the end in case of failure?

Reply:
That's correct; I'll figure out a better name, since it IS referencing the `bestAttemptContent` attribute in the extension.